### PR TITLE
Connection management for SFMC

### PR DESCRIPTION
**Pull request by bharatgulati**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -56,7 +56,7 @@
     <surefire.redirectTestOutputToFile>true</surefire.redirectTestOutputToFile>
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
     <!-- version properties -->
-    <cdap.version>6.8.0</cdap.version>
+    <cdap.version>6.8.0-SNAPSHOT</cdap.version>
```
Comment (rmstar) [inline pom.xml]: why are you depending on SNAPSHOT version?

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+  public ETApiObject row;
+  private Iterator<? extends ETApiObject> iterator;
+
+  MarketingConnector (MarketingConnectorConfig config) {
+  this.config = config;
+  }
+
+  @Override
+  public void test (ConnectorContext connectorContext) throws ValidationException {
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java]: remove space after method name. test ( -> test(. Same comment in other places

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+  FailureCollector collector = connectorContext.getFailureCollector();
+  config.validateCredentials(collector);
+  }
+
+  @Override
+  public BrowseDetail browse (ConnectorContext connectorContext, BrowseRequest browseRequest) throws IOException {
+  BrowseDetail.Builder browseDetailBuilder = BrowseDetail.builder();
+  int count = 0;
+      for (SourceObject object : SourceObject.values()) {
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java]: fix indent

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+  @Override
+  public void test (ConnectorContext connectorContext) throws ValidationException {
+  FailureCollector collector = connectorContext.getFailureCollector();
+  config.validateCredentials(collector);
+  }
+
+  @Override
+  public BrowseDetail browse (ConnectorContext connectorContext, BrowseRequest browseRequest) throws IOException {
+  BrowseDetail.Builder browseDetailBuilder = BrowseDetail.builder();
+  int count = 0;
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java]: return value seems to be hardcoded - independent of the input.. why is that?

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java (hunk 3 above)
Comment (rmstar) [reply src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java]: indent is off in multiple places. please make sure to fix indent in all files

File: docs/MarketingCloud-batchsource.md [hunk 5]
```diff
@@ -92,9 +92,6 @@ For example, `https://instance.auth.marketingcloudapis.com/`
 **SOAP API Endpoint**: The SOAP Endpoint URL associated for the Server-to-Server API integration. For example, 
 `https://instance.soap.marketingcloudapis.com/Service.asmx`
 
-**REST API Base URI**: The REST API Base URL associated for the Server-to-Server API integration. For example,
```
Comment (rmstar) [inline docs/MarketingCloud-batchsource.md:95]: why was this removed?

File: src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudSource.java [hunk 6]
```diff
@@ -57,7 +56,6 @@
 @Description("Read marketing data from Salesforce Marketing cloud.")
 public class MarketingCloudSource extends BatchSource<NullWritable, StructuredRecord, StructuredRecord> {
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudSource.java:59]: annotate with @Metadata

**Review by rmstar — COMMENTED** [review]
Took a quick first look. Please ask Rahul to review

File: pom.xml (hunk 1 above)
Comment (bharatgulati) [reply pom.xml]: Earlier I was facing the issue with 6.8.0 version with ETL API  package which was showing it does not exist. Now i have resolved the issue and changed it to cdap version 6.8.0

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java (hunk 2 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java]: Removed the space after method test

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java (hunk 3 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java]: Done. I have fixed indent for all new classes.

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java (hunk 4 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java]: Okay, i have removed the count variable and changed it to - browseDetailBuilder.setTotalCount(SourceObject.values().length).build();

File: docs/MarketingCloud-batchsource.md (hunk 5 above)
Comment (bharatgulati) [reply docs/MarketingCloud-batchsource.md:95]: It was not being used in sink. That's why I had removed it.

File: src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudSource.java (hunk 6 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudSource.java:59]: Done

File: pom.xml (hunk 1 above)
Comment (rmstar) [reply pom.xml]: change to 6.8.0

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java (hunk 3 above)
Comment (rmstar) [reply src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java]: still not fixed. maybe you forgot to push a commit?

File: pom.xml (hunk 1 above)
Comment (bharatgulati) [reply pom.xml]: Done. Changed it to 6.8.0

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java (hunk 3 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java]: Yes, Now I have pushed a commit.

**Comment by MrRahulSharma** [comment]
Correct the indentation in all the files.

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnectorConfig.java [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+              .withConfigProperty(MarketingCloudConstants.PROPERTY_CLIENT_ID)
+              .withConfigProperty(MarketingCloudConstants.PROPERTY_CLIENT_SECRET)
+              .withConfigProperty(MarketingCloudConstants.PROPERTY_AUTH_API_ENDPOINT)
+              .withConfigProperty(MarketingCloudConstants.PROPERTY_SOAP_API_ENDPOINT)
+              .withStacktrace(e.getStackTrace());
+    }
+  }
+
+  /**
+   * Returns true if Salesforce can be connected to.
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnectorConfig.java]: Correct the Javadoc comment : Returns true if any of the required parameters are macro enabled.

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+import java.util.Map;
+
+/**
+ * Salesforce Marketing Cloud Connector
+ */
+@Plugin(type = Connector.PLUGIN_TYPE)
+@Name(MarketingCloudConstants.PLUGIN_NAME)
+@Description("Connection to access data from Salesforce Marketing Objects.")
+@Metadata
+public class MarketingConnector implements DirectConnector {
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java]: Rename to MarketingCloudConnector.

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+       properties)).addRelatedPlugin(new PluginSpec(MarketingCloudDataExtensionSink.PLUGIN_TYPE,
+       BatchSink.PLUGIN_TYPE, properties)).build();
+  }
+
+  @Override
+  public List<StructuredRecord> sample (ConnectorContext connectorContext, SampleRequest sampleRequest)
+          throws IOException {
+    String objectName = sampleRequest.getPath();
+    if (objectName == null) {
+      throw new IllegalArgumentException("Path should contain object name");
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java]: "Path should not be null." is enough.

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+  }
+    try {
+      return listObjectDetails(SourceObject.valueOf(objectName));
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java]: Use SourceObject.fromValue method instead of SourceObject.valueOf method at all places.

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+  } catch (ETSdkException e) {
+      throw new IOException("unable to fetch records", e);
+  }
+  }
+
+  private List<StructuredRecord> listObjectDetails (SourceObject sourceObject) throws ETSdkException
+          , IOException {
+    List<StructuredRecord> sampleList = new ArrayList<>();
+    MarketingCloudUtil cloudUtil = new MarketingCloudUtil();
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java]: Restructure this method properly. Something like

List<StructuredRecord> sampleList = new ArrayList<>();
//  returning EmptyList for DATA_EXTENSION Source Object because we need to provide requestId which is dynamic. However, provided requestId null.
if (SourceObject.DATA_EXTENSION != sourceObject) {
MarketingCloudUtil cloudUtil = new MarketingCloudUtil();
MarketingCloudClient marketingCloudClient = MarketingCloudClient.create
          (config.getClientId(), config.getClientSecret(), config.getAuthEndpoint(), config.getSoapEndpoint());
response = marketingCloudClient.fetchObjectRecords(sourceObject, null);
      MarketingCloudObjectInfo sfObjectMetaData = marketingCloudClient.fetchObjectSchema(sourceObject);
      iterator = response.getObjects().iterator();
      while (iterator.hasNext()) {
          StructuredRecord.Builder builder = StructuredRecord.builder(schema);
          row = iterator.next();
          cloudUtil.convertRecord(sfObjectMetaData, builder, row);
          sampleList.add(builder.build());
      }
}
return sampleList;

File: src/main/java/io/cdap/plugin/sfmc/source/util/MarketingCloudUtil.java [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+import org.slf4j.LoggerFactory;
+import java.lang.reflect.Method;
+import java.util.Date;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Salesforce Marketing Cloud Conversion
+ */
+public class MarketingCloudUtil {
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/sfmc/source/util/MarketingCloudUtil.java:36]: The instance creation of this class is not required, make this a static class instead.

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnectorConfig.java [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+      !containsMacro(MarketingCloudConstants.PROPERTY_SOAP_API_ENDPOINT) &&
+      !containsMacro(MarketingCloudConstants.PROPERTY_AUTH_API_ENDPOINT);
+  }
+
+  public Schema getSchema(SourceObject sourceObject) {
+    try {
+      MarketingCloudClient client = MarketingCloudClient.create(this.getClientId(),
+        this.getClientSecret(),
+        this.getAuthEndpoint(),
+        this.getSoapEndpoint());
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnectorConfig.java]: This might be an heavy object to create as internally it makes a bunch of calls to create the ETClient object which contains the access token and REST/SOAP connection objects. Can you check if we can optimise on the number of instances required of MarketingCloudClient?

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+  private List<StructuredRecord> listObjectDetails(SourceObject sourceObject) throws ETSdkException
+    , IOException {
+    List<StructuredRecord> sampleList = new ArrayList<>();
+    MarketingCloudUtil cloudUtil = new MarketingCloudUtil();
+    MarketingCloudClient marketingCloudClient = MarketingCloudClient.create
+      (config.getClientId(), config.getClientSecret(), config.getAuthEndpoint(), config.getSoapEndpoint());
+    //  returning Collections.emptyList for DATA_EXTENSION Source Object because dataExtensionKey is dynamic and
+    //  to fetch the records for DATA_EXTENSION object, dataExtensionKey is needed and it can't be hardcoded here.
+    if (SourceObject.DATA_EXTENSION == sourceObject) {
+      return Collections.emptyList();
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java]: Check with the existing Salesforce clients for concluding if we cannot support displaying all DATA_EXTENSION objects. If we cannot then please add a note in the connector documentation to capture this limitation.

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnectorConfig.java (hunk 7 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnectorConfig.java]: Changed it to "Returns true if any of the required parameters are macro enabled."

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java (hunk 8 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java]: Changed the class name from MarketingConnector to MarketingCloudConnector.

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java (hunk 9 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java]: Changed it to "Path should not be null."

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java (hunk 10 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java]: SourceObject.fromValue method will not work because it is expecting name of the object e.g. 'Data Extension' but we are forming URL based on its enum value which is DATA_EXTENSION.
However, currently using SourceObject.valueOf method.

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java (hunk 11 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java]: I have restructured the code after adding the logic for fetching Customer keys of DATA_EXTENSION object.

File: src/main/java/io/cdap/plugin/sfmc/source/util/MarketingCloudUtil.java (hunk 12 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/plugin/sfmc/source/util/MarketingCloudUtil.java:36]: Done.

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnectorConfig.java (hunk 13 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnectorConfig.java]: Created a function 'getInstance' in MarketingCloudUtil class and accessing the function wherever it is required to avoid multiple calls to create ETClient object.

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java (hunk 14 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnector.java]: Added the logic to fetch list of Customer keys and their records of DATA_EXTENSION Object. Now it is evident on the UI while browsing.

File: src/main/java/io/cdap/plugin/sfmc/source/util/MarketingCloudUtil.java [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+           return convertToValue(fieldName, fieldSchema.getNonNullable(), fieldValue);
+       }
+   throw new IllegalStateException(
+           String.format("Field '%s' is of unexpected type '%s'. Declared 'complex UNION' types: %s",
+                   fieldName, fieldValue.getClass().getSimpleName(), fieldSchema.getUnionSchemas()));
+   default:
+       throw new IllegalStateException(
+               String.format("Record type '%s' is not supported for field '%s'", fieldType.name(), fieldName));
+  }
+  }
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/sfmc/source/util/MarketingCloudUtil.java:113]: Correct the indentation and line spacings in this file. Also do ensure new lines between function definitions as well.

File: src/main/java/io/cdap/plugin/sfmc/source/util/SourceObject.java [hunk 16]
```diff
@@ -36,11 +36,6 @@
  */
 public enum SourceObject {
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/sfmc/source/util/SourceObject.java]: The changes in this file is not required and can be reverted completely.

File: src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudSourceConfig.java [hunk 17]
```diff
@@ -148,35 +121,26 @@ public class MarketingCloudSourceConfig extends PluginConfig {
...
    * @param soapEndpoint      The SOAP API endpoint for Salesforce Marketing Cloud
    */
   public MarketingCloudSourceConfig(String referenceName, String queryMode, @Nullable String objectName,
                                     @Nullable String dataExtensionKey, @Nullable String objectList,
                                     @Nullable String dataExtensionKeys, @Nullable String tableNameField,
                                     @Nullable String filter, String clientId, String clientSecret,
-                                    String restEndpoint, String authEndpoint, String soapEndpoint) {
-    this.referenceName = referenceName;
+                                    String authEndpoint, String soapEndpoint)
+    throws ETSdkException {
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudSourceConfig.java]: Please check if we require to have this throws statement.

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingCloudConnector.java [hunk 18]
```diff
@@ -0,0 +1,255 @@
...
+
+  public static ETClient create(String clientId, String clientSecret, String authEndpoint,
+                                String soapEndpoint) throws ETSdkException {
+    ETConfiguration conf = new ETConfiguration();
+    conf.set("clientId", clientId);
+    conf.set("clientSecret", clientSecret);
+    conf.set("authEndpoint", authEndpoint);
+    conf.set("soapEndpoint", soapEndpoint);
+    conf.set("useOAuth2Authentication", "true");
+    conf.set("cxfConnectTimeout", "300000");
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/sfmc/connector/MarketingCloudConnector.java]: Use constants for connect and receive timeouts.

File: src/main/java/io/cdap/plugin/sfmc/sink/MarketingCloudConf.java [hunk 19]
```diff
@@ -112,30 +91,27 @@ public class MarketingCloudConf extends PluginConfig {
...
-
-  String getClientSecret() {
-    return clientSecret;
+  @VisibleForTesting
+  public MarketingCloudConf(String referenceName, String clientId, String clientSecret, String dataExtension,
+                            String authEndpoint,
+                            String soapEndpoint, String restEndPoint, @Nullable Integer maxBatchSize,
+                            @Nullable String operation,
+                            @Nullable String columnMapping, Boolean failOnError, Boolean replaceWithSpaces,
+                            Boolean truncateText) throws ETSdkException {
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/sfmc/sink/MarketingCloudConf.java]: Can we check if we really require this throws statement.

File: src/main/java/io/cdap/plugin/sfmc/source/util/MarketingCloudUtil.java [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+import java.lang.reflect.Method;
+import java.util.Date;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Salesforce Marketing Cloud Conversion
+ */
+public class MarketingCloudUtil {
+  private static final Logger LOG = LoggerFactory.getLogger(MarketingConnectorConfig.class);
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/sfmc/source/util/MarketingCloudUtil.java]: Possible typo: Should be MarketingCloudUtil.class

File: src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudJobConfiguration.java
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudJobConfiguration.java:1]: Nit: The change in the file is not required.

File: src/main/java/io/cdap/plugin/sfmc/source/util/MarketingCloudUtil.java (hunk 12 above)
Comment (MrRahulSharma) [reply src/main/java/io/cdap/plugin/sfmc/source/util/MarketingCloudUtil.java:36]: Can you update the PR with this change ?

File: src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudJobConfiguration.java
Comment (bharatgulati) [reply src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudJobConfiguration.java:1]: Sure.

File: src/main/java/io/cdap/plugin/sfmc/source/util/MarketingCloudUtil.java (hunk 20 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/plugin/sfmc/source/util/MarketingCloudUtil.java]: changed it to MarketingCloudUtil.class

File: src/main/java/io/cdap/plugin/sfmc/sink/MarketingCloudConf.java (hunk 19 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/plugin/sfmc/sink/MarketingCloudConf.java]: Removed throws statement.

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingCloudConnector.java (hunk 18 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/plugin/sfmc/connector/MarketingCloudConnector.java]: Using constants instead of passing like this "300000"

File: src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudSourceConfig.java (hunk 17 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/plugin/sfmc/source/MarketingCloudSourceConfig.java]: Removed this unused throws statement.

File: src/main/java/io/cdap/plugin/sfmc/source/util/SourceObject.java (hunk 16 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/plugin/sfmc/source/util/SourceObject.java]: Reverted back completely.

File: src/main/java/io/cdap/plugin/sfmc/source/util/MarketingCloudUtil.java (hunk 15 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/plugin/sfmc/source/util/MarketingCloudUtil.java:113]: Fixed the indentation and line spacing.

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingCloudConnector.java [hunk 21]
```diff
@@ -78,6 +78,8 @@ public class MarketingCloudConnector implements DirectConnector {
   private Schema schema;
   private Iterator<? extends ETApiObject> iterator;
 
+  private static String timeOut = "300000";
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/sfmc/connector/MarketingCloudConnector.java]: Use 'private static **final**', and rename to CXF_TIMEOUT_VALUE

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingCloudConnector.java [hunk 22]
```diff
@@ -0,0 +1,257 @@
...
+
+  @Override
+  public List<StructuredRecord> sample(ConnectorContext connectorContext, SampleRequest sampleRequest)
+    throws IOException {
+    MarketingCloudClient marketingCloudClient = null;
+    MarketingCloudUtil cloudUtil = new MarketingCloudUtil();
+    try {
+      marketingCloudClient = MarketingCloudUtil.getInstance(config.getClientId(),
+        config.getClientSecret(), config.getAuthEndpoint(), config.getSoapEndpoint());
+    String objectName = sampleRequest.getPath();
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/sfmc/connector/MarketingCloudConnector.java]: Correct the line indentations in all the files.

File: src/main/java/io/cdap/plugin/sfmc/source/util/MarketingCloudUtil.java (hunk 15 above)
Comment (MrRahulSharma) [reply src/main/java/io/cdap/plugin/sfmc/source/util/MarketingCloudUtil.java:113]: The indentation is still not right at many places. PTAL

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingCloudConnector.java (hunk 21 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/plugin/sfmc/connector/MarketingCloudConnector.java]: Renamed to 'CXF_TIMEOUT_VALUE' and Using private static final.

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingCloudConnector.java (hunk 22 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/plugin/sfmc/connector/MarketingCloudConnector.java]: Fixed the line indentation in all the files. Please have a look.

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnectorConfig.java [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+  public boolean shouldConnect() {
+    return !containsMacro(MarketingCloudConstants.PROPERTY_CLIENT_ID) &&
+      !containsMacro(MarketingCloudConstants.PROPERTY_CLIENT_SECRET) &&
+      !containsMacro(MarketingCloudConstants.PROPERTY_SOAP_API_ENDPOINT) &&
+      !containsMacro(MarketingCloudConstants.PROPERTY_AUTH_API_ENDPOINT);
+  }
+
+  public Schema getSchema(SourceObject sourceObject) throws ETSdkException {
+    return MarketingCloudInputFormat.getTableMetaData(sourceObject, null, MarketingCloudUtil.getInstance
+        (getClientId(), getClientSecret(), getAuthEndpoint(), getSoapEndpoint())).getSchema();
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnectorConfig.java]: getTableMetaData returns null in case of exception, please handle null case properly.

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnectorConfig.java (hunk 13 above)
Comment (MrRahulSharma) [reply src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnectorConfig.java]: MarketingCloudUtil is not the correct place for this method. Instead keep the method in the MarketingCloudClient and rename it to getOrCreate.

File: src/main/java/io/cdap/plugin/sfmc/source/util/MarketingCloudUtil.java (hunk 15 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/plugin/sfmc/source/util/MarketingCloudUtil.java:113]: Fixed the indentation

File: src/main/java/io/cdap/plugin/sfmc/source/util/SourceObject.java (hunk 16 above)
Comment (MrRahulSharma) [reply src/main/java/io/cdap/plugin/sfmc/source/util/SourceObject.java]: Please revert this file.

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnectorConfig.java (hunk 23 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnectorConfig.java]: Done

File: src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnectorConfig.java (hunk 13 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnectorConfig.java]: moved it to MarketingCloudClient and changed the function name to getOrCreate.

**Comment by MrRahulSharma** [comment]
Please squash the commits into one.

**Review by MrRahulSharma — APPROVED** [review]

File: src/test/java/io/cdap/plugin/sfmc/sink/RecordDataExtensionRowConverterTest.java [hunk 24]
```diff
@@ -110,5 +110,105 @@ public void testTransform() throws ETSdkException {
...
+    StructuredRecord structuredRecord = Mockito.mock(StructuredRecord.class);
+    Mockito.when(structuredRecord.getSchema()).thenReturn(inputSchema);
+    Mockito.when(structuredRecord.get("emailid")).thenReturn("abc@123");
+    LocalDate date = LocalDate.of(2011, 01, 01);
+    Mockito.when(structuredRecord.get("abcd")).thenReturn(date);
+    Mockito.when(structuredRecord.getDate("abcd")).thenReturn(date);
+    Mockito.when(structuredRecord.getDecimal("price")).thenReturn(BigDecimal.valueOf(2000));
+    List<String> list = new ArrayList<>();
+    list.add("abcd");
+    try {
```
Comment (MrRahulSharma) [inline src/test/java/io/cdap/plugin/sfmc/sink/RecordDataExtensionRowConverterTest.java]: Ideally Assert statements should not be put inside the try-catch block. Please make sure to move all the asserts outside the try-catch. In case we do expect an exception to be thrown make that as a new test case which repros the exception scenario and expects and exception in the process.

File: src/test/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReaderTest.java [hunk 25]
```diff
@@ -183,7 +148,12 @@ public void testFetchData() throws Exception {
     Mockito.doReturn(response).when(client).fetchObjectRecords(object, null);
     Mockito.doReturn(results).when(response).getObjects();
     marketingCloudRecordReader.initialize(split, null);
-    Assert.assertTrue(marketingCloudRecordReader.nextKeyValue());
+    try {
+      marketingCloudRecordReader.nextKeyValue();
+      collector.getOrThrowException();
+    } catch (IOException e) {
+      Assert.assertEquals(79, e.getMessage().length());
```
Comment (Aryan-Verma) [inline src/test/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReaderTest.java]: Why are we comparing the exception message length here? Can we just compare the actual message?

File: src/test/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReaderTest.java [hunk 26]
```diff
@@ -227,38 +197,154 @@ public void testFetchDataWithDataExtension() throws Exception {
...
 
   @Test
-  public void testGetCurrentKeyValue() throws IOException, NoSuchFieldException {
+  public void testGetCurrentKeyValue() throws NoSuchFieldException {
     ETApiObject row = new ETDataExtensionRow();
     row.setId("id");
     List<MarketingCloudColumn> columns = new ArrayList<>();
     MarketingCloudColumn column = new MarketingCloudColumn("name", "String");
     columns.add(column);
+    MarketingCloudRecordReader marketingCloudRecordReader1 = new MarketingCloudRecordReader(marketingCloudSourceConfig);
```
Comment (Aryan-Verma) [inline src/test/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReaderTest.java]: nit: terminology

Please change it to something like newMarketingCloudRecordReader or marketingCloudRecordReaderCopy

File: src/test/java/io/cdap/plugin/sfmc/sink/RecordDataExtensionRowConverterTest.java (hunk 24 above)
Comment (bharatgulati) [reply src/test/java/io/cdap/plugin/sfmc/sink/RecordDataExtensionRowConverterTest.java]: Yes, In those test cases we do expect an exception to be thrown. However, I have done the changes as per the need with exception approach. Added comments as well. Please have a look.

File: src/test/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReaderTest.java (hunk 25 above)
Comment (bharatgulati) [reply src/test/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReaderTest.java]: Done. Changed it to the actual message instead of length.

File: src/test/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReaderTest.java (hunk 26 above)
Comment (bharatgulati) [reply src/test/java/io/cdap/plugin/sfmc/source/MarketingCloudRecordReaderTest.java]: Done. Renamed it to 'cloudRecordReader'.

**Review by Aryan-Verma — APPROVED** [review]
